### PR TITLE
Add RouteGroup to allow registration of middleware consistently

### PR DESCRIPTION
**Pull request by opengeek**
Requires routes be collected and compiled before usage; see changes to App->run() and App->__invoke()

Attempts to address #1265

### CODE REVIEW

File: Slim/Interfaces/RouterInterface.php [hunk 1]
```diff
@@ -42,12 +43,11 @@ public function dispatch(ServerRequestInterface $request);
...
      * Add a route group to the array
      *
-     * @param string     $group      The group pattern prefix
-     * @param array|null $middleware Optional middleware
+     * @param RouteGroup &$group The route group
      *
      * @return int The index of the new group
      */
-    public function pushGroup($group, $middleware = []);
+    public function pushGroup(&$group);
```
Comment (JoeBengalen) [inline Slim/Interfaces/RouterInterface.php]: Why the reference sign? As it seems you are passing in an object, which will always be passed by reference.

Also maybe typehint for `RouteGroup`

File: Slim/App.php [hunk 2]
```diff
@@ -233,7 +234,7 @@ public function map(array $methods, $pattern, $callable)
             $callable = $callable->bindTo($this);
         }
 
-        $route = $this->container->get('router')->map($methods, $pattern, $callable);
+        $route = $this->router->map($methods, $pattern, $callable);
```
Comment (akrabat) [inline Slim/App.php]: Why make this change?

File: Slim/RouteGroup.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        return $this;
+    }
+
+    /**
+     * Invoke the group to register any Routable objects within it.
+     *
+     * @param App &$app The App to bind the callable to.
+     */
+    function __invoke(&$app)
```
Comment (JoeBengalen) [inline Slim/RouteGroup.php]: Same here; No reference needed and typehint for `App`

File: Slim/Interfaces/RouterInterface.php (hunk 1 above)
Comment (akrabat) [inline Slim/Interfaces/RouterInterface.php]: Also why do we allow an array of middleware?

**Comment by silentworks** [comment]
Can the creation of Group's happen inside of the router instead of in the App please?

File: Slim/Interfaces/RouterInterface.php (hunk 1 above)
Comment (opengeek) [inline Slim/Interfaces/RouterInterface.php]: Reference sign was leftover from some silly attempts to get add() to work on the already pushed groups (snicker), never mind that they are popped immediately after evaluation. :(

The array of middleware was an artifact of allowing any number of arguments in the call to `group()`, but this is gone now with `add()` working.

File: Slim/App.php (hunk 2 above)
Comment (opengeek) [inline Slim/App.php]: Probably so I could see the methods available on the router which cannot be inferred from the `container->get()` call. I'll put it back...

File: Slim/RouteGroup.php (hunk 3 above)
Comment (opengeek) [inline Slim/RouteGroup.php]: Noted and reverting.

**Comment by opengeek** [comment]
@silentworks I'm not sure how you envision the change to move the logic inside the Router. That would then require that I pass the App to a similar method inside Router. Is that really an improvement? It feels like I'm just moving the same dependency around, but if you think that is preferable I can make that change. Or if you have another suggestion?

**Comment by opengeek** [comment]
See https://gist.github.com/opengeek/c8a2f8952560e226855b for example usage

**Pull request merged**